### ISSUE TITLE
Extension causes high cpu load

### ISSUE THREAD

**Issue by KikeVen** [microsoft/pylance-release#2983]
- Issue Type: `Performance`
- Extension Name: `vscode-pylance`
- Extension Version: `2022.6.30`
- OS Version: `Windows_NT x64 10.0.19044`
- VS Code version: `1.69.0-insider`

:warning: Make sure to **attach** this file from your *home*-directory:
:warning:`file:///c%3A/[ms-python.vscode-pylance-unresponsive.cpuprofile.txt](https://github.com/microsoft/pylance-release/files/9022080/ms-python.vscode-pylance-unresponsive.cpuprofile.txt)`

Find more details here: https://github.com/microsoft/vscode/wiki/Explain-extension-causes-high-cpu-load

**Comment by bschnurr** [comment]
can you provide the logs? https://github.com/microsoft/pylance-release/blob/main/TROUBLESHOOTING.md#filing-an-issue

is there a public repo to test on?

**Comment by KikeVen** [comment]
No there isn’t, thank you

**Comment by millerdev** [comment]
I'm seeing the same. Two cores pegged at 100%

![image](https://user-images.githubusercontent.com/464726/177412188-a7ccebec-b863-416e-9733-17364c4b6744.png)

This started happening today, I assume just after my system automatically upgraded to 2022.6.30.

**Comment by millerdev** [comment]
> Which version of Python do you use,

Python 3.9.12 (the patch version should not matter)

> and which platform(s)?

I run Linux, but other devs on the team run MacOS. I do not know if anyone on MacOS experienced the issue.

> I would normally suggest turning off symbol indexing to see if that might be the problem, but I'm guessing you've already tried that.

I have not. I'll give it a try today and get back to you.

> Perhaps you could try manually downgrading to the previous version to confirm that the problem is isolated to the latest version?

I'll also give that a try if turning off symbol indexing does nothing.

Thanks for the prompt response.

**Comment by millerdev** [comment]
Reproduction steps:

Checkout this branch: [`dm/sql-lookup-tables`](https://github.com/dimagi/commcare-hq/compare/dm/sql-lookup-tables)

Apply the edits in this patch:

```diff
diff --git a/corehq/apps/fixtures/couchmodels.py b/corehq/apps/fixtures/couchmodels.py
index f5da1aa7191..ca8b65ad873 100644
--- a/corehq/apps/fixtures/couchmodels.py
+++ b/corehq/apps/fixtures/couchmodels.py
@@ -318,12 +318,6 @@ class FixtureDataItem(SyncCouchToSQLMixin, Document):
             return {key: value.to_api_json()
                     for key, value in self.fields.items()}
 
-    @property
-    def data_type(self):
-        if not hasattr(self, '_data_type'):
-            self._data_type = FixtureDataType.get(self.data_type_id)
-        return self._data_type
-
     def add_owner(self, owner, owner_type, transaction=None):
         assert(owner.domain == self.domain)
         with transaction or CouchTransaction() as transaction:
@@ -376,18 +370,6 @@ class FixtureDataItem(SyncCouchToSQLMixin, Document):
     def remove_location(self, location):
         return self.remove_owner(location, 'location')
 
-    def type_check(self):
-        fields = set(self.fields.keys())
-        for field in self.data_type.fields:
-            if field.field_name in fields:
-                fields.remove(field)
-            else:
-                raise FixtureTypeCheckError(
-                    f"field {field.field_name} not in fixture data {self.get_id}")
-        if fields:
-            raise FixtureTypeCheckError(
-                f"fields {', '.join(fields)} from fixture data {self.get_id} not in fixture data type")
-
     def get_groups(self, wrap=True):
         group_ids = get_owner_ids_by_type(self.domain, 'group', self.get_id)
         if wrap:
```

Scroll to the top of the file and delete the unreferenced import:

```py
    FixtureTypeCheckError,
```

Note that the line below it (`FixtureVersionError`) now appears to be unreferenced even though it's not. Undo to restore `FixtureTypeCheckError`. Here's what I see (comments added):

![image](https://user-images.githubusercontent.com/464726/177548457-697a14a5-fcc2-4aa7-b189-cad3896a2635.png)

Hover over `FixtureVersionError`, note the incorrect warning: `"FixtureTypeCheckError" is not accessed Pylance`

![image](https://user-images.githubusercontent.com/464726/177551277-8f7994fd-936c-486b-8aee-c7af7d466326.png)

Wait a while (20-30 seconds) for the references to resolve. All the while I have two cores pegged at 100%.

Trace log from that editing session:
[pls-trace.log](https://github.com/microsoft/pylance-release/files/9054734/pls-trace.log)

**Comment by millerdev** [comment]
> I would normally suggest turning off symbol indexing to see if that might be the problem...

Gave this a try, and it improved things, so I turned it back on to see if the issue would come back. Performance also was good at first (after turning it on again), but after a while it is now back to two cores pegged at 100%.

**Comment by heejaechang** [comment]
looks like this file is the problem - (657261) [BG(1)] analyzing: /home/dimagi/code/commcare-hq/corehq/apps/fixtures/views.py (104045ms)

took almost 2 min analyzing.

**Comment by erictraut** [comment]
Yes, I'm investigating now. I have it narrowed down to a specific method, but I still don't understand the root cause. I'll update this thread once I have an update.

**Comment by millerdev** [comment]
I rolled back to 2022.6.20 for ~6 hours and observed no issues with CPU usage or sluggish linting. The pegged cores returned minutes after upgrading to 2022.6.30 again.

**Comment by erictraut** [comment]
@millerdev, thanks for the repro steps and logs. This has been really helpful.

I was able to track down the specific change in pyright that led to the increase in analysis time. It was an innocuous-looking bug fix that I would have never expected to have this effect. It had a subtle and difficult-to-explain interaction with the rest of the type evaluation engine. There are cases where type evaluation involves exponential-time operations (`O(x^n)`); the trick is to keep `x` small and `n` as bounded as possible. In your code sample, the bug fix caused `n` to grow slightly, which caused the entire operation to shoot up in terms of analysis time.

I've introduced a fix that restores the previous performance for your code. So worst case, it will now work as well as it did prior to 2022.6.30. I plan to spend some time looking for additional optimizations because I'm still not satisfied with the performance in this case.

if you're curious, the method that pyright is struggling with is `_update_items` in `apps/fixtures/views.py`. And in particular, it's struggling to infer the type of the variable `fields_json`, which involves assignments where the assigned expression depends on the symbol being assigned within a doubly-nested loop. If you temporarily comment out the statement `fields_json = JsonSet(fields_json, [new_field_name], [])`, the type analysis for this source file completes almost immediately. I still need to do more exploration to understand the root cause.

**Comment by erictraut** [comment]
For anyone coming to this issue in the future, please note that almost every "high CPU load" issue is different, so if you are experiencing this symptom, please open a new issue with logs and repro steps rather than adding to this issue.

**Comment by erictraut** [comment]
It was introduced in pyright 1.1.255. I think pylance 2022.6.30 was the first version of pylance that picked it up.

**Comment by erictraut** [comment]
I figured out the root cause of the performance issue. In case you are interested in the details...

When Pyright encounters a class that is not generic (i.e. it doesn't derive from `typing.Generic` or other generic base classes) and has a constructor with one or more unannotated parameters, it treats the class as though it is generic and is parameterized by n type variables, one for each constructor input parameter. This allows for better type inference of instance variables initialized from the constructor parameters. For example:

```python
class Foo:
    def __init__(self, a, b):
        self.a = a
        self.b = b

x = Foo(1, "hi")
reveal_type(x.a) # int
reveal_type(x.b) # str

x = Foo(1.2, 4j)
reveal_type(x.a) # float
reveal_type(x.b) # complex
```

This mechanism works well in most cases, but in your code it led to a performance problem when evaluating the type of a variable named `fields_json` in the function `_update_items`. This variable has no type annotation, so its type must be inferred from assigned values. It is assigned within a doubly-nested loop, and the variable has a dependency on itself. It is assigned an expression that includes a call to the `JsonSet` constructor. The first time through the loop, the type analyzer inferred the type `JsonSet[str, Unknown]`. The next time through the loop, it inferred `JsonSet[str, Unknown] | JsonSet[JsonSet[str, Unknown], Unknown]`, and this recursion continued until the type checker hit an internal limiter designed to prevent infinite recursion. The limiter did its job at preventing a crash, but it took a long time before the limiter was hit. Because of the doubly-nested loop, this limiter needed to be hit n^2 times before the type was fully resolved.

I've added a fix that detects this case and handles it without the need for any iteration. The entire file can now be analyzed in tens of milliseconds rather than multiple seconds. The fix will be included in the next release.

Thanks again for providing the details repro steps.
